vault/token_store.go

```diff
@@ -1444,11 +1446,26 @@ func (ts *TokenStore) revokeTreeInternal(ctx context.Context, id string) error {
 		}
 
 		path := saltedID + "/"
-		children, err := ts.parentView(saltedNS).List(saltedCtx, path)
+		childrenRaw, err := ts.parentView(saltedNS).List(saltedCtx, path)
 		if err != nil {
 			return errwrap.Wrapf("failed to scan for children: {{err}}", err)
 		}
 
+		// Filter the child list to remove any items that have ever been in the dfs stack.
+		// This is a robustness check, as a parent/child cycle can lead to an OOM crash.
+		children := make([]string, 0, len(childrenRaw))
+		for _, child := range childrenRaw {
+			if _, seen := seenIDs[child]; !seen {
+				children = append(children, child)
+			} else {
+				if err = ts.parentView(saltedNS).Delete(saltedCtx, path+child); err != nil {
```

Comment:
`ts.revokeInternal` down below should be making this exact `Delete` call on the child token, so I wonder if the cycle is is due to the token not being properly cleaned up in there. IMO `revokeTreeInternal` should be only in charge of tree traversal and delegate deletion to `revokeInternal` to ensure that all related entries are properly removed.

Reply:
`ts.revokeInternal` does make this delete for the parent reference in the token, but if we're here, there are extraneous parent references that will never get deleted. e.g. assume the state is:
```
parent/a/c
parent/b/c
```

If `c`'s correct parent is `a`, that info is stored in `c` and will cause `parent/a/c` to be deleted in `revokeInternal`. But nothing would trigger deletion of `parent/b/c`.

A prior version of the PR didn't have the delete step. It worked fine and was resilient to cycles, but @briankassouf noted that it would leave stray parent references.

Reply:
I don't think it can ever get to that state though right? `c` can only ever have one parent, unless entry creation was borked at some point.

Reply:
Root cause for how a cycle could be created is still unknown, but we did see it in user data (see: https://hashicorp.slack.com/archives/C04L34UGZ/p1536860851000100) and if present, the revocation will just consume memory until Vault crashes.